File: python/Factory/pizza_store.py
from abc import ABC, abstractmethod
from pizza import CheesePizza, ClamPizza, Pizza
from pizza_ingredient_factory import ChicagoPizzaIngredientFactory, NYPizzaIngredientFactory, PizzaIngredientFactory


class PizzaStore(ABC):
    @abstractmethod
    def __init__(self) -> None:
        self.ingredient_factory: PizzaIngredientFactory

    def order_pizza(self, type: str) -> Pizza:
        pizza: Pizza = self.create_pizza(type)
        pizza.prepare()
        pizza.bake()
        pizza.cut()
        pizza.box()
        return pizza

    @abstractmethod
    def create_pizza(self, type: str) -> Pizza:
        pass


class NYPizzaStore(PizzaStore):
    def __init__(self) -> None:
        self.ingredient_factory: PizzaIngredientFactory = NYPizzaIngredientFactory()

    def create_pizza(self, type: str) -> Pizza:
        if type == "cheese":
            name = "NY Style Sauce and Cheese Pizza"
            return CheesePizza(name, self.ingredient_factory)
        elif type == "clam":
            name = "NY Style Clam Pizza"
            return ClamPizza(name, self.ingredient_factory)
        else:
            raise Exception("this type do not exist")


class ChicagoPizzaStore(PizzaStore):
    def __init__(self) -> None:
        self.ingredient_factory: PizzaIngredientFactory = ChicagoPizzaIngredientFactory()

    def create_pizza(self, type: str) -> Pizza:
        if type == "cheese":
            name = "Chicago Style Deep Dish Cheese Pizza"
            return CheesePizza(name, self.ingredient_factory)
        else:
            raise Exception("this type do not exist")

File: c/Factory/pizza_store.c
#include "pizza_store.h"

void PizzaStoreNew(PizzaStore *pstore)
{
    PizzaStoreInherit(pstore);
}
Pizza *PizzaStoreOrderPizza(PizzaStore *pstore, char *type)
{
    Pizza *pizza = pstore->createPizza(pstore, type);
    pizza->prepare(pizza);
    pizza->bake(pizza);
    pizza->cut(pizza);
    pizza->box(pizza);
    return pizza;
}

void NYPizzaStoreNew(NYPizzaStore *ny_pstore)
{
    NYPizzaIngredientFactoryNew(&ny_pstore->ingredient_factory);
    PizzaStoreInherit(ny_pstore);
    ny_pstore->createPizza = NYPizzaStoreCreatePizza;
}
Pizza *NYPizzaStoreCreatePizza(PizzaStore *ny_pstore, char *type)
{
    if (strcmp(type, "cheese") == 0)
    {
        CheesePizza *pizza = (CheesePizza *)malloc(sizeof(CheesePizza));
        CheesePizzaNew(pizza, &ny_pstore->ingredient_factory);
        strcpy(pizza->name, "NY Style Sauce and Cheese Pizza");
        return (Pizza *)pizza;
    }
    else if (strcmp(type, "clam") == 0)
    {
        ClamPizza *pizza = (ClamPizza *)malloc(sizeof(ClamPizza));
        ClamPizzaNew(pizza, &ny_pstore->ingredient_factory);
        strcpy(pizza->name, "NY Style Clam Pizza");
        return (Pizza *)pizza;
    }
    else
    {
        printf("this type do not exist");
    }
    return 0;
}

void ChicagoPizzaStoreNew(ChicagoPizzaStore *ch_pstore)
{
    ChicagoPizzaIngredientFactoryNew(&ch_pstore->ingredient_factory);
    PizzaStoreInherit(ch_pstore);
    ch_pstore->createPizza = ChicagoPizzaStoreCreatePizza;
}
Pizza *ChicagoPizzaStoreCreatePizza(PizzaStore *ch_pstore, char *type)
{
    if (strcmp(type, "cheese") == 0)
    {
        CheesePizza *pizza = (CheesePizza *)malloc(sizeof(CheesePizza));
        CheesePizzaNew(pizza, &ch_pstore->ingredient_factory);
        strcpy(pizza->name, "Chicago Style Deep Dish Cheese Pizza");
        return (Pizza *)pizza;
    }
    else
    {
        printf("this type do not exist");
    }
    return 0;
}